pkg/clusteragent/clusterchecks/dispatcher_rebalance.go

```diff
@@ -0,0 +1,208 @@
+		node.RLock()
+		busyness := calculateBusyness(node.clcRunnerStats)
+		node.RUnlock()
+		diffMap[nodeName] = busyness - avg
+		weights = append(weights, Weight{
+			nodeName: nodeName,
+			busyness: busyness,
+		})
+	}
+	return diffMap, weights
+}
+
+// updateDiff creates a map that contains the difference between
+// the busyness on each node and the total average busyness.
+func (d *dispatcher) updateDiff(avg int) map[string]int {
+	diffMap := make(map[string]int)
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for nodeName, node := range d.store.nodes {
+		node.RLock()
+		busyness := calculateBusyness(node.clcRunnerStats)
+		node.RUnlock()
+		diffMap[nodeName] = busyness - avg
+	}
+	return diffMap
+}
+
+// pickCheckToMove select the most appropriate check to move from a node to another.
+// A check Xi running on a node N is chosen to move to another node if it satisfies the following
+// Weight(Xi) >  Weight(Xj) (for each j != i, 0 <= j < len(weights))
+// where Weight(X) is the busyness value caused by running the check X.
+func (d *dispatcher) pickCheckToMove(nodeName string) (string, int, error) {
+	(*d.store.nodes[nodeName]).RLock()
+	defer (*d.store.nodes[nodeName]).RUnlock()
+
+	if len((*d.store.nodes[nodeName]).clcRunnerStats) == 0 {
+		return "", -1, fmt.Errorf("Node %s has no check stats", nodeName)
+	}
+
+	firstItr := true
+	checkID := ""
+	checkWeight := 0
+	for id, stats := range (*d.store.nodes[nodeName]).clcRunnerStats {
+		busyness := int(busynessFunc(stats.AverageExecutionTime, stats.MetricSamples))
+		if busyness > checkWeight || firstItr {
+			checkWeight = busyness
+			checkID = id
+			firstItr = false
+		}
+	}
+	return checkID, checkWeight, nil
+}
+
+// pickNode select the most appropriate node to receive a specific check.
+// A node Ni is most appropriate to receive a check with a weight W
+// if it satisfies the following
+// Diff(Ni) < Diff(Nj) (for each j != i, 0 <= j < len(nodes))
+// where Diff(N) is the difference between the busyness on N and the total average busyness.
+func pickNode(diffMap map[string]int, srcNode string) string {
+	firstItr := true
+	minDiff := 0
+	pickedNode := ""
+	for _, node := range orderedKeys(diffMap) {
+		if node == srcNode {
+			continue
+		}
+		if diffMap[node] < minDiff || firstItr {
+			minDiff = diffMap[node]
+			pickedNode = node
+			firstItr = false
+		}
+	}
+	return pickedNode
+}
+
+// moveCheck moves a check by its ID from a node to another
+func (d *dispatcher) moveCheck(src, dest, checkID string) {
+	log.Debugf("Moving %s from %s to %s", checkID, src, dest)
+
+	d.store.RLock()
+	destNode, destFound := d.store.getNodeStore(dest)
+	srcNode, srcFound := d.store.getNodeStore(src)
+	d.store.RUnlock()
+
+	if !destFound || !srcFound {
+		log.Debugf("Nodes not found in store: %s, %s. Check %s will not move ", src, dest, checkID)
+		return
+	}
+
+	srcNode.Lock()
+	destNode.Lock()
+
+	runnerStats := srcNode.getRunnerStats(checkID)
+	destNode.addRunnerStats(checkID, runnerStats)
```

Comment:
in the same logic that I proposed in other comments, the `Lock()` should be done in the `addRunnerStats(....)` method